core/src/main/java/org/apache/iceberg/CatalogUtil.java

```diff
@@ -171,45 +172,57 @@ public static Catalog loadCatalog(
 
   /**
    * Load a custom {@link FileIO} implementation.
+   */
+  public static FileIO loadFileIO(
+      String impl,
+      Map<String, String> properties,
+      Configuration hadoopConf) {
+    return loadCatalogConfigurable(impl, properties, hadoopConf, FileIO.class);
+  }
+
+  /**
+   * Load a custom implementation of a {@link CatalogConfigurable}.
    * <p>
    * The implementation must have a no-arg constructor.
-   * If the class implements {@link Configurable},
+   * If the class implements Hadoop's {@link Configurable},
    * a Hadoop config will be passed using {@link Configurable#setConf(Configuration)}.
-   * {@link FileIO#initialize(Map properties)} is called to complete the initialization.
+   * {@link CatalogConfigurable#initialize(Map properties)} is called to complete the initialization.
    *
    * @param impl full class name of a custom FileIO implementation
    * @param hadoopConf hadoop configuration
+   * @param resultClass the final return class type
    * @return FileIO class
    * @throws IllegalArgumentException if class path not found or
    *  right constructor not found or
    *  the loaded class cannot be casted to the given interface type
    */
-  public static FileIO loadFileIO(
+  public static <T extends CatalogConfigurable> T loadCatalogConfigurable(
       String impl,
       Map<String, String> properties,
-      Configuration hadoopConf) {
-    LOG.info("Loading custom FileIO implementation: {}", impl);
-    DynConstructors.Ctor<FileIO> ctor;
+      Configuration hadoopConf,
+      Class<T> resultClass) {
+    LOG.info("Loading custom {} implementation: {}", resultClass.getName(), impl);
+    DynConstructors.Ctor<T> ctor;
     try {
-      ctor = DynConstructors.builder(FileIO.class).impl(impl).buildChecked();
+      ctor = DynConstructors.builder(resultClass).impl(impl).buildChecked();
```

Comment:
Looks like the `class` in `DynConstructors.builder(Class)` is only used during hiddenImpl. Can probably just use `DynConstructors.builder()` unless you want to keep that open for later?

Reply:
I'm also wondering if we should just use a ServiceLoader instead of using the dyn constructors stuff? It seems to be more heavy duty

Reply:
it's also a little ridiculous having to serialize all the config for FileIO and others into Map<String, String> so that it can be loaded from properties if it's already there.

Is there some way Spark could support a Map<String, Serializable> which can get passed through to DataSourceV2? Is that already getting discussed somewhere?

Reply:
> I'm also wondering if we should just use a ServiceLoader instead of using the dyn constructors stuff?

I don't think this would be a good choice. `ServiceLoader` will scan the entire classpath for matching classes, which takes a really long time in large applications for initialization. It also introduces runtime issues because service files need to be merged when creating shaded Jars and it is easy to overlook.

Since we already know the class to load from configuration that is passed to the catalog, I don't think `ServiceLoader` is a good option compared to using reflection to load a class. And the Dyn* classes just help make that easier to read and maintain.

Reply:
> it's also a little ridiculous having to serialize all the config for FileIO and others into Map<String, String> so that it can be loaded from properties if it's already serializable.

I don't think I'm following what you are suggesting. The configuration passed in comes from a string map, like a Spark config or Hadoop config. Serialization is a concern after catalogs are configured because the components that were created on the driver need to be serialized to be sent for use on executors. Those are separate enough that I'm not sure how to improve on this.

Reply:
> it's also a little ridiculous having to serialize all the config for FileIO and others into Map<String, String> so that it can be loaded from properties if it's already serializable.
> ```

Yes as Ryan said, I don't think it serializes all the config. `Map<String, String> properties` is passed in during Spark initialization time and only initialized once. For AWS module, `AwsProperties` is serialized, but it is a much smaller subset of all the configurations.

Reply:
You're right, serialization is unrelated. I'm probably missing stuff, but my understanding of this PR is that it's trying to mimic being able to pass an aws service client builder through to the catalog?

But since the catalog only can get passed Map<String, String>, what you do is pass a custom builder class name and a flat set of args. Then when it wants the builder, it will instantiate your builder class and pass you back the set of args which you're expected to put back into the right places.

I think an example is if I want the catalog to use this config for it's dynamo client:
```
Dynamo.builder()
  .withRegion("myRegion")
  .withProxyConfig(
    ProxyConfig
    .withPort(1000)
    .withHost("localhost")
    .build())
  .build()
```
This pr is letting a user create this:
```
CustomDynamoBuilder {
   public Dynamo dynamo;
   public void initialize(Map<String, String> props) {
     dynamo = Dynamo.builder()
.withRegion(props.get("region")
.withProxyConfig(ProxyConfig
  .withPort(PropertyUtils.propertyAsInt(port, "port"))
  .withHost(props.get("host"))
  .build();
   }
}
```
Then the user is expected to push through into the source: Map(dynamoClass -> CustomDynamoBuilder, region -> myRegion, port -> 1000, host -> localhost).

So the user will have to reimplement the aws builders for every service they need to customize by making a mapping from a flat string list to every property in the config.

Is this basically that, except that it's more modularized by letting you define a ProxyConfigBuilder which will then try to update each custom builder you make with that proxy config if it's allowed (and other things like proxies aka credentials)?

Reply:
I don't think the user needs to reimplement the aws builders for every service, see the example I provided in the conversation below. In the v2 AWS client, the following aspects can be override:
1. credential provider
2. region
3. endpoint (url + port)
4. http client (for things like proxy)
5. client configuration
6. service configuration

we can either provide a way to dynamically configure each type of AWS client, or dynamically configure each aspect of the clients. From my current observation it is more common for users to want to use the same configuration across all clients, but maybe also have some exceptions for some specific clients. So it seems to make more sense to go with this proposed approach.

For your given example, yes information like port needs to be passed in, but even in a normal application it needs to be passed in from some configuration file. In this method you can also read your own configuration system instead of reading from the catalog properties passed in.

You can read the example I gave below for more information, and there is no need to initialize the client from the user.

Reply:
@jackye1995 for sure. It's awesome that's it's so configurable. I've been pulling apart our codebase to try to strip out s3a (which is way harder to configure on the version I'm on than this) and I'm excited for when I can test this out! 

`it's also a little ridiculous having to serialize all the config` this line was worded poorly and was meant to be sympathetic about all the config mapping you're having to add, not the implementation. (It was after hours of trying to translate between my own internal configuration)